bluemira/builders/tf_coils.py

```diff
@@ -0,0 +1,337 @@
+        """
+        Make a single BioSavart Filament for a single TF coil
+        """
+        bb = self.wp_cross_section.bounding_box
+        dx_xs = 0.5 * (bb.x_max - bb.x_min)
+        dy_xs = 0.5 * (bb.y_max - bb.y_min)
+
+        dx_wp, dy_wp = [0], [0]  # default to coil centreline
+        if self.nx > 1:
+            dx_wp = np.linspace(
+                dx_xs * (1 / self.nx - 1), dx_xs * (1 - 1 / self.nx), self.nx
+            )
+
+        if self.ny > 1:
+            dy_wp = np.linspace(
+                dy_xs * (1 / self.ny - 1), dy_xs * (1 - 1 / self.ny), self.ny
+            )
+
+        current_wires = []
+        for dx in dx_wp:
+            c_wire = offset_wire(wire, dx)
+            for dy in dy_wp:
+                c_w = deepcopy(c_wire)
+                c_w.translate((0, dy, 0))
+                current_wires.append(c_w)
+
+        current_arrays = [
+            w.discretize(byedges=True, dl=wire.length / 200) for w in current_wires
+        ]
+
+        # We need all arrays to be CCW, this will hopefully go away with a fix for #482
+        from bluemira.geometry._deprecated_tools import check_ccw
+
+        for c in current_arrays:
+            if not check_ccw(c[:, 0], c[:, 2]):
+                c[:, 0] = c[:, 0][::-1]
+                c[:, 1] = c[:, 1][::-1]
+                c[:, 2] = c[:, 2][::-1]
+
+        radius = 0.5 * BluemiraFace(self.wp_cross_section).area / (self.nx * self.ny)
+        filament = BiotSavartFilament(
+            current_arrays, radius=radius, current=1 / (self.nx * self.ny)
+        )
+        return filament
+
+    def update_cage(self, x):
+        """
+        Update the magnetostatic solver
+        """
+        super().update_parameterisation(x)
+        wire = self.parameterisation.create_shape()
+        circuit = self._make_single_circuit(wire)
+
+        self.cage = HelmholtzCage(circuit, self.params.n_TF.value)
+        field = self.cage.field(self.params.R_0, 0, self.params.z_0)
+        current = -self.params.B_0 / field[1]  # single coil amp-turns
+        current /= self.nx * self.ny  # single filament amp-turns
+        self.cage.set_current(current)
+
+    def calculate_ripple(self, x):
+        """
+        Calculate the ripple on the target points for a given variable vector
+        """
+        self.update_cage(x)
+        ripple = self.cage.ripple(*self.ripple_points)
+        self.ripple_values = ripple
+        return ripple - self.params.TF_ripple_limit
+
+    def f_constrain_ripple(self, constraint, x, grad):
+        """
+        Toroidal field ripple constraint function
+        """
+        constraint[:] = self.calculate_ripple(x)
+
+        if grad.size > 0:
+            # Only called if a gradient-based optimiser is used
+            grad[:] = self.optimiser.approx_derivative(
+                self.calculate_ripple, x, constraint
+            )
+        bluemira_debug(f"Max ripple: {max(constraint+self.params.TF_ripple_limit)}")
```

Comment:
I think we should have a debug flush logger otherwise the terminal gets quite messy (it will all be printed in the log file anyway)

Reply:
Will do, not sure how many flush logs we'll end up with (I guess it makes less sense for warnings and above), so I'll probs just put a new version of the existing function in that supports debug styling rather than generalising any further.